File: services/banking-service/src/main/java/zw/co/microbank/bankingservice/service/AccountService.java
package zw.co.microbank.bankingservice.service;

import zw.co.microbank.bankingservice.dto.AccountResponse;
import zw.co.microbank.bankingservice.dto.CreateAccountRequest;

import java.math.BigDecimal;
import java.util.List;

/**
 * Author: tjc
 * Created on 7/10/25
 */

public interface AccountService {
    //    Create Account
    AccountResponse createAccount(CreateAccountRequest request);

    // Get an account by client id
    AccountResponse getAccountByClientId(String clientId);

    //    Get an account by id
    AccountResponse getAccountById(String id);

    //    Get an account by account number
    AccountResponse getAccountByAccountNumber(String accountNumber);

    //    Get all accounts
    List<AccountResponse> getAllAccounts();

    //    Get account balance
    BigDecimal getAccountBalance(String accountNumber);
}

File: client/src/services/account.service.ts
import {
    AccountResponseDto,
    CreateAccountRequestDto,
} from '@/dtos/account.dto';
import {getStandardHeaders, handleApiResponse} from "@/services/api.utils";


const API_BASE_URL = process.env.NEXT_PUBLIC_BACKEND_API_URL || 'http://localhost:8080';
const ACCOUNTS_ENDPOINT = `${API_BASE_URL}/api/v1/banks/accounts`;


export const accountService = {

    async createAccount(request: CreateAccountRequestDto): Promise<AccountResponseDto> {
        const response = await fetch(ACCOUNTS_ENDPOINT, {
            method: 'POST',
            headers: getStandardHeaders(true),
            body: JSON.stringify(request),
        });
        return handleApiResponse<AccountResponseDto>(response);
    },

    async getAccountByClientId(clientId: string): Promise<AccountResponseDto> {
        const response = await fetch(`${ACCOUNTS_ENDPOINT}/client/${clientId}`, {
            method: 'GET',
            headers: getStandardHeaders(true),
        });
        return handleApiResponse<AccountResponseDto>(response);
    },

    async getAccountById(id: string): Promise<AccountResponseDto> {
        const response = await fetch(`${ACCOUNTS_ENDPOINT}/${id}`, {
            method: 'GET',
            headers: getStandardHeaders(true),
        });
        return handleApiResponse<AccountResponseDto>(response);
    },

    async getAccountByAccountNumber(accountNumber: string): Promise<AccountResponseDto> {
        const response = await fetch(`${ACCOUNTS_ENDPOINT}/number/${accountNumber}`, {
            method: 'GET',
            headers: getStandardHeaders(true),
        });
        return handleApiResponse<AccountResponseDto>(response);
    },

    async getAllAccounts(): Promise<AccountResponseDto[]> {
        const response = await fetch(ACCOUNTS_ENDPOINT, {
            method: 'GET',
            headers: getStandardHeaders(true),
        });
        return handleApiResponse<AccountResponseDto[]>(response);
    },

    async getAccountBalance(accountNumber: string): Promise<string> {
        const response = await fetch(`${ACCOUNTS_ENDPOINT}/balance/${accountNumber}`, {
            method: 'GET',
            headers: getStandardHeaders(true),
        });

        return handleApiResponse<string>(response);
    },
};